### ISSUE TITLE
Early stopping is implemented incorrectly in XGBoost.java:203 (jvm-packages)

### ISSUE THREAD

**Issue by yvirin-apixio** [dmlc/xgboost#3140]
## Environment info

`xgboost` version used: 0.8
 
The commit hash (`git rev-parse HEAD`):
11bfa8584d8ddba2060ae2731f8599184e65f3b9

## The description of the problem
In the file: https://github.com/dmlc/xgboost/blob/master/jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/XGBoost.java line 203:

seems like the "&=" is not the correct operator to use, and instead it should be "|=" with decreasing initialized to false, in order to make sure consecutive "increases" in the loss function during evaluation rounds have been observed. With the current implementation the code stops after first increase, which is wrong. Probably it is a hard bug to catch in practice, I was just inspecting the implementation of early stopping and found this issue.
The intended usage of earlyStoppingRound parameter is described in line 121 (comment).

**Comment by superbobry** [comment]
Feel free to submit a PR fixing this. There is also #2982.

**Comment by yvirin-apixio** [comment]
@superbobry Thanks for connecting to the other issue. Somewhat related question: currently there is no way to know which is the optimal treeLimit to use in predictions, correct? Presumably it would be the tree which performed the best on evaluation during training.

**Comment by hcho3** [comment]
Closing this for now. Feel free to re-open for any updates.
